### PR TITLE
fix(embeddings): silent NULL embeddings after marketplace upgrades

### PR DESCRIPTION
**Pull request by efenocchi**
## Summary

Embeddings have been silently failing in production. The capture hook
spawns the embed daemon from the marketplace plugin bundle path
(`~/.claude/plugins/cache/hivemind/hivemind/<version>/bundle/embeddings/embed-daemon.js`),
but `@huggingface/transformers` is installed only at
`~/.hivemind/embed-deps/node_modules/` — reachable from the bundle via
a one-time symlink that `hivemind embeddings install` creates. Every
Claude Code marketplace auto-upgrade drops a new versioned cache dir
without the symlink, so the bundle daemon's
`import("@huggingface/transformers")` returns MODULE_NOT_FOUND,
`embed()` returns null, and `sessions.message_embedding` writes through
as `NULL` forever after. The user-facing log shows ~30 minutes of
`embed err: Cannot find package` lines.

This PR is a fix-it-forever overhaul of the embed daemon lifecycle:

1. **Explicit transformers resolver in the daemon.** `nomic.ts` now
   loads transformers via
   `createRequire(pathToFileURL("~/.hivemind/embed-deps/")).resolve()`
   followed by an absolute-URL dynamic import, with a bare-specifier
   fallback for dev trees. The daemon resolves transformers regardless
   of which bundle path it was spawned from.

2. **Persistent opt-in via `~/.deeplake/config.json`.** The legacy
   `HIVEMIND_EMBEDDINGS` env var is read ONCE on first run to seed
   `embeddings.enabled`, then never consulted again. Source of truth
   is now a real file shared across all 4 agents.

3. **New CLI surface.** `hivemind embeddings install` is the heavy
   path (deps + symlinks + `enabled:true`). `enable` / `disable` are
   light config flips. `uninstall [--prune]` is the heavy counterpart
   (removes symlinks + optionally wipes shared deps). Every agent's
   SessionStart injection advertises the full surface.

4. **Hello handshake + stuck-daemon recycle.** The socket is per-UID,
   not per-version, so one stuck daemon poisons every subsequent
   session for up to 10 minutes of idle-out. The client now sends a
   `hello` on first connect; on path mismatch (typical post-upgrade)
   it SIGTERMs the daemon and clears its sock/pid. Same recycle fires
   if the daemon returns a transformers-missing error during
   `embed()`. Next session spawns fresh from the current bundle.

5. **Visible one-time notification.** On transformers-missing, the
   client enqueues a warn-severity notification routed through the
   existing SessionStart drain so the user sees a clear "run
   `hivemind embeddings install`" message. Suppressed when the user
   has explicitly disabled embeddings — no nag for opt-outs.

6. **Self-heal symlink across cached versions.** Each capture hook
   runs `ensurePluginNodeModulesLink({ bundleDir })` at top-level.
   The first capture under a new plugin version (post-marketplace-
   upgrade) creates `<pluginDir>/node_modules → ~/.hivemind/embed-deps/node_modules`
   atomically; subsequent calls are O(1) no-ops. Conservative — never
   clobbers a real `node_modules`, never overrides a symlink to a
   different valid target, but cleans up dangling links so it can
   recreate them on the next call. Refuses to act outside the
   production `bundle/` layout (guard against tests).

7. **Bundle-scan guards.** New `tests/claude-code/embeddings-bundle-scan.test.ts`
   reads each agent's shipped `embed-daemon.js`, `capture.js`, and
   the CLI bundle, and asserts every literal string the runtime
   depends on survives bundling. 30-second reviewer guardrail.

## Behavioral migration

| Was | Now |
|---|---|
| `HIVEMIND_EMBEDDINGS=false` env var | `~/.deeplake/config.json` → `embeddings.enabled: false` |
| `EmbeddingsStatus = "env-disabled"` | `EmbeddingsStatus = "user-disabled"` |
| `enableEmbeddings()` (heavy) | `installEmbeddings()` |
| `disableEmbeddings({ prune })` (heavy) | `uninstallEmbeddings({ prune })` |
| (n/a) | `enableEmbeddings()` (light) — flip config flag, warn if deps missing |
| (n/a) | `disableEmbeddings()` (light) — flip flag + SIGTERM daemon |
| First-run migration | env-var read exactly once: `false`/unset → `enabled: false`; else `enabled: true` |

Most users have the env var unset, so on first run after this lands
they'll get `enabled: false` and need to run `hivemind embeddings enable`
or `install` to opt back in. That's intentional — explicit consent
instead of implicit on-by-default.

## Test plan

- [x] `npm test` → 2457 passing (up from ~2371 on `main`).
- [x] `tsc --noEmit` clean.
- [x] Bundle scan asserts every agent's shipped artifact contains the
  fix.
- [x] Self-heal helper handles every fs combination (missing target,
  existing real dir, existing symlink-elsewhere, dangling symlink,
  non-bundle layout).
- [ ] **Manual sandbox capture test (per CLAUDE.md "never touch prod
  tables")**:
  - Switch to test org:
    `node codex/bundle/commands/auth-login.js org switch test_plugin`
  - Point `sessionsTableName` at `sessions_test` in `src/config.ts`,
    rebuild, reinstall plugin.
  - Kill any pre-existing daemon:
    `kill $(cat /tmp/hivemind-embed-$(id -u).pid 2>/dev/null) ;
     rm -f /tmp/hivemind-embed-$(id -u).{sock,pid}`
  - Drive a real Claude Code session for a few prompts.
  - `SELECT COUNT(*) AS total, COUNT(message_embedding) AS with_embed
    FROM sessions_test
    WHERE creation_date > NOW() - INTERVAL '5 minutes';`
  - Expect `with_embed == total`.
  - Revert `src/config.ts` before merging.
- [ ] Manual disable/enable path:
  - `hivemind embeddings disable` → next capture writes NULL, no
    daemon spawn, no notification.
  - `hivemind embeddings enable` → daemon spawns, embedding column
    populated.
- [ ] Manual self-heal:
  - `rm <marketplace-cache>/<version>/node_modules`
  - Next capture event → symlink recreated.

## Out of scope (filed separately)

- 403 retries from the Deeplake SQL endpoint visible in
  `~/.deeplake/hook-debug.log`. Different auth-layer issue,
  orthogonal to embeddings.

## Summary by CodeRabbit

* **New Features**
  * Added `hivemind embeddings install` to manage transformer dependencies separately from enablement
  * Added `hivemind embeddings uninstall [--prune]` for full cleanup with optional dependency pruning
  * Embeddings configuration now persists in `~/.deeplake/config.json` for cross-session preference preservation
  * Daemon verification on startup to detect stale bundles and auto-recovery
  * User notifications when required dependencies are missing

* **Bug Fixes**
  * Improved handling of stuck or incompatible embedding daemons with automatic recycling
  * Better error detection and messaging for missing transformer dependencies

### CODE REVIEW

File: bundle/cli.js [hunk 1]
```diff
@@ -3668,49 +3763,104 @@ function ensureSharedDeps() {
...
     chmodSync(SHARED_DAEMON_PATH, 493);
   } else {
     warn(`  Embeddings     standalone daemon bundle missing at ${src} (run 'npm run build' first)`);
   }
 }
 function linkAgent(install) {
-  const link = join10(install.pluginDir, "node_modules");
+  const link = join11(install.pluginDir, "node_modules");
   symlinkForce(SHARED_NODE_MODULES, link);
   log(`  Embeddings     linked ${install.id.padEnd(20)} -> shared deps`);
```
[the remark this answers is not in the document]
Comment (efenocchi) [reply bundle/cli.js:3891]: Fixed in 5b56452 — `linkAgent()` now calls `linkStateFor()` first and skips the symlink-replacement path when the existing entry is a real directory (the same `owns-own-node-modules` state surfaced in `status`). Test in tests/cli/cli-embeddings.test.ts asserts the real dir + marker file are preserved across the install call.

File: bundle/cli.js [hunk 2]
```diff
@@ -3668,49 +3763,104 @@ function ensureSharedDeps() {
...
+  const sockPath = socketPathFor(String(uid));
+  let pid = null;
+  try {
+    pid = Number.parseInt(readFileSync8(pidPath, "utf-8").trim(), 10);
+    if (Number.isFinite(pid)) {
+      try {
+        process.kill(pid, "SIGTERM");
+      } catch {
+      }
+    }
```
[the remark this answers is not in the document]
Comment (efenocchi) [reply bundle/cli.js]: Fixed in f406030 — `killEmbedDaemon()` now probes the UDS path with a 200ms timeout (`_isDaemonAliveOnSocket`) before SIGTERMing. If the socket is dead the daemon is gone, the PID may have been recycled, so we skip the kill and only clean up sock/pid files. Test writes the current process PID into the pidfile, deletes the socket, and asserts the test runner is NOT killed.

File: claude-code/bundle/capture.js [hunk 3]
```diff
@@ -1469,59 +1743,90 @@ function embeddingSqlLiteral(vec) {
...
+    try {
+      statSync(link);
+      return { kind: "linked-elsewhere", link, existingTarget };
+    } catch {
+      try {
+        rmSync(link);
+      } catch {
+      }
+      return { kind: "stale-link-removed", link, danglingTarget: existingTarget };
+    }
```
[the remark this answers is not in the document]
Comment (efenocchi) [reply claude-code/bundle/capture.js:1949]: Fixed in dd1ce08 (same fix covers #13) — `ensurePluginNodeModulesLink()` now falls through to `createSymlinkAtomic()` immediately after rm'ing the dangling link, so the link is repaired in the same call. Test asserts `existsSync(link) === true` and `readlinkSync` points at the shared target after a single invocation.

File: claude-code/bundle/pre-tool-use.js [hunk 4]
```diff
@@ -1073,13 +1073,171 @@ function pidPathFor(uid, dir = DEFAULT_SOCKET_DIR) {
...
+  }
+  mkdirSync2(join4(home, ".deeplake"), { recursive: true, mode: 448 });
+  const tmp = `${path}.${process.pid}.tmp`;
+  writeFileSync2(tmp, JSON.stringify(q, null, 2), { mode: 384 });
+  renameSync(tmp, path);
+}
+function enqueueNotification(n) {
+  const q = readQueue();
+  q.queue.push(n);
+  writeQueue(q);
```
[the remark this answers is not in the document]
Comment (efenocchi) [reply claude-code/bundle/pre-tool-use.js]: Fixed in f2f7895 — `enqueueNotification()` now runs under `withQueueLock()`, an O_EXCL `.lock` file with retry-with-backoff and a 5s stale-lock reclaim. Test spawns 12 parallel subprocesses each enqueueing one unique notification; asserts the final queue has exactly 12 entries (was 1-3 sequentially before the lock).

File: claude-code/bundle/pre-tool-use.js [hunk 5]
```diff
@@ -1134,6 +1305,106 @@ var EmbedClient = class {
...
+      return;
+    }
+    if (hello.daemonPath === this.daemonEntry)
+      return;
+    if (_recycledStuckDaemon)
+      return;
+    _recycledStuckDaemon = true;
+    log4(`daemon path mismatch \u2014 running=${hello.daemonPath} expected=${this.daemonEntry}; recycling`);
+    this.recycleDaemon(hello.pid);
+  }
```
[the remark this answers is not in the document]
Comment (efenocchi) [reply claude-code/bundle/pre-tool-use.js:1456]: Fixed in 19a911f — `verifyDaemonOnce()` no longer sets `helloVerified = true` before awaiting the probe. The flag is set ONLY after a compatible hello response; probe failures and recycle branches leave it false so the next reconnect retries. Test simulates a transient probe failure (server destroys socket without responding to the first hello) and asserts `helloAttempts === 2` across three `embed()` calls.

File: claude-code/bundle/shell/deeplake-shell.js [hunk 6]
```diff
@@ -68050,7 +68288,7 @@ function normalizeSessionMessage(path2, message) {
   return normalizeContent(path2, raw);
 }
 function resolveEmbedDaemonPath() {
-  return join9(dirname4(fileURLToPath(import.meta.url)), "embeddings", "embed-daemon.js");
+  return join11(dirname5(fileURLToPath(import.meta.url)), "embeddings", "embed-daemon.js");
 }
```
[the remark this answers is not in the document]
Comment (efenocchi) [reply claude-code/bundle/shell/deeplake-shell.js:68416]: Fixed in 4b1ebb8 (same fix covers #7 and #11) — `src/shell/deeplake-fs.ts` resolver was building `dirname(import.meta.url) + 'embeddings/embed-daemon.js'`. The shell bundle lives at `<agent>/bundle/shell/`, so without `..` it pointed at the missing `<agent>/bundle/shell/embeddings/embed-daemon.js`. Fix adds the `..`. Test asserts the daemon file exists at the parent-of-shell location AND the literal `'..'` survives bundling immediately before `'embeddings'`.

File: codex/bundle/shell/deeplake-shell.js [hunk 7]
```diff
@@ -68050,7 +68288,7 @@ function normalizeSessionMessage(path2, message) {
   return normalizeContent(path2, raw);
 }
 function resolveEmbedDaemonPath() {
-  return join9(dirname4(fileURLToPath(import.meta.url)), "embeddings", "embed-daemon.js");
+  return join11(dirname5(fileURLToPath(import.meta.url)), "embeddings", "embed-daemon.js");
 }
```
[the remark this answers is not in the document]
Comment (efenocchi) [reply codex/bundle/shell/deeplake-shell.js:68416]: Fixed in 4b1ebb8 — same fix as the claude-code shell bundle above. The source resolver in `src/shell/deeplake-fs.ts` is shared across all agent bundles; one-line addition of `..` repairs codex/cursor/claude-code in lockstep.

File: codex/bundle/stop.js [hunk 8]
```diff
@@ -1149,13 +1149,171 @@ function pidPathFor(uid, dir = DEFAULT_SOCKET_DIR) {
...
+  }
+  mkdirSync8(join13(home, ".deeplake"), { recursive: true, mode: 448 });
+  const tmp = `${path}.${process.pid}.tmp`;
+  writeFileSync7(tmp, JSON.stringify(q, null, 2), { mode: 384 });
+  renameSync4(tmp, path);
+}
+function enqueueNotification(n) {
+  const q = readQueue();
+  q.queue.push(n);
+  writeQueue(q);
```
[the remark this answers is not in the document]
Comment (efenocchi) [reply codex/bundle/stop.js]: Fixed in 0796d90 (same fix covers #12) — cross-process dedup is now done in `enqueueNotification()` itself: before appending, it checks whether a notification with the same `(id, dedupKey)` is already in the queue and skips if so. Test runs 3 sequential subprocesses with identical `(id, dedupKey)` and asserts the queue has exactly 1 entry.

File: cursor/bundle/shell/deeplake-shell.js [hunk 9]
```diff
@@ -67741,17 +67907,22 @@ var EmbedClient = class {
...
+      await this.verifyDaemonOnce(sock);
       const id = String(++this.nextId);
       const req = { op: "embed", id, kind, text };
       const resp = await this.sendAndWait(sock, req);
       if (resp.error || !("embedding" in resp) || !resp.embedding) {
-        log3(`embed err: ${resp.error ?? "no embedding"}`);
+        const err = resp.error ?? "no embedding";
+        log4(`embed err: ${err}`);
+        if (isTransformersMissingError(err)) {
+          this.handleTransformersMissing(err);
```
[the remark this answers is not in the document]
Comment (efenocchi) [reply cursor/bundle/shell/deeplake-shell.js]: Fixed in 2ca402b — `embed()` is split into a thin outer wrapper plus `embedAttempt()`; when `verifyDaemonOnce()` returns the new `"recycled"` sentinel, the wrapper spawns a fresh daemon and retries once. With autoSpawn off (tests / pi's canonical-shared-deps fallback) it just returns null so the call doesn't stall on a dead socket. Test asserts `embed_attempts === 0` after recycle (request never sent on the dead connection).

File: cursor/bundle/shell/deeplake-shell.js [hunk 10]
```diff
@@ -67908,6 +68179,9 @@ var EmbedClient = class {
 function sleep2(ms3) {
   return new Promise((r10) => setTimeout(r10, ms3));
 }
+function isTransformersMissingError(err) {
+  return /(@huggingface\/transformers|hivemind embeddings install|MODULE_NOT_FOUND)/i.test(err);
+}
```
[the remark this answers is not in the document]
Comment (efenocchi) [reply cursor/bundle/shell/deeplake-shell.js:68308]: Fixed in e74b854 (same fix covers #14) — `isTransformersMissingError()` no longer matches bare `MODULE_NOT_FOUND`. The matcher now requires either the actionable-wrapper string (`hivemind embeddings install`) or a direct `@huggingface/transformers` reference. Negative tests assert `MODULE_NOT_FOUND` for `onnxruntime-node` / `sharp` are NOT classified as transformers issues.

File: cursor/bundle/shell/deeplake-shell.js [hunk 11]
```diff
@@ -68050,7 +68288,7 @@ function normalizeSessionMessage(path2, message) {
   return normalizeContent(path2, raw);
 }
 function resolveEmbedDaemonPath() {
-  return join9(dirname4(fileURLToPath(import.meta.url)), "embeddings", "embed-daemon.js");
+  return join11(dirname5(fileURLToPath(import.meta.url)), "embeddings", "embed-daemon.js");
 }
```
[the remark this answers is not in the document]
Comment (efenocchi) [reply cursor/bundle/shell/deeplake-shell.js:68416]: Fixed in 4b1ebb8 — same shared resolver fix as the claude-code/codex shell bundles. Adding `..` in `src/shell/deeplake-fs.ts` flows into the cursor bundle on rebuild.

File: hermes/bundle/capture.js [hunk 12]
```diff
@@ -579,13 +579,171 @@ function pidPathFor(uid, dir = DEFAULT_SOCKET_DIR) {
...
+  mkdirSync2(join4(home, ".deeplake"), { recursive: true, mode: 448 });
+  const tmp = `${path}.${process.pid}.tmp`;
+  writeFileSync2(tmp, JSON.stringify(q, null, 2), { mode: 384 });
+  renameSync(tmp, path);
+}
+function enqueueNotification(n) {
+  const q = readQueue();
+  q.queue.push(n);
+  writeQueue(q);
+}
```
[the remark this answers is not in the document]
Comment (efenocchi) [reply hermes/bundle/capture.js]: Fixed in 0796d90 — same fix as #8: cross-process dedup at the queue layer. `_signalledMissingDeps` is in-process only; the `(id, dedupKey)` collapse in `enqueueNotification()` now handles cross-process duplicates regardless of which agent's hook produced them.

File: hermes/bundle/capture.js [hunk 13]
```diff
@@ -802,78 +1076,103 @@ function embeddingSqlLiteral(vec) {
...
+    try {
+      statSync(link);
+      return { kind: "linked-elsewhere", link, existingTarget };
+    } catch {
+      try {
+        rmSync(link);
+      } catch {
+      }
+      return { kind: "stale-link-removed", link, danglingTarget: existingTarget };
+    }
```
[the remark this answers is not in the document]
Comment (efenocchi) [reply hermes/bundle/capture.js:1246]: Fixed in dd1ce08 — same fix as #3: `ensurePluginNodeModulesLink()` rm's the dangling link AND immediately recreates it. The hermes bundle picks this up on rebuild since the self-heal source is shared. Caller no longer needs to ignore the `stale-link-removed` return value to get a working bundle.

File: hermes/bundle/wiki-worker.js [hunk 14]
```diff
@@ -376,41 +647,8 @@ var EmbedClient = class {
...
-function embeddingsStatus() {
-  if (cachedStatus !== null)
-    return cachedStatus;
-  cachedStatus = detectStatus();
-  return cachedStatus;
-}
-function embeddingsDisabled() {
-  return embeddingsStatus() !== "enabled";
+function isTransformersMissingError(err) {
+  return /(@huggingface\/transformers|hivemind embeddings install|MODULE_NOT_FOUND)/i.test(err);
```
[the remark this answers is not in the document]
Comment (efenocchi) [reply hermes/bundle/wiki-worker.js]: Fixed in e74b854 — same matcher narrow as #10. Bundles regenerated for all agents including hermes wiki-worker.

File: src/embeddings/client.ts [hunk 15]
```diff
@@ -99,6 +125,94 @@ export class EmbedClient {
...
+      try {
+        pid = Number.parseInt(readFileSync(this.pidPath, "utf-8").trim(), 10);
+      } catch { /* no pidfile */ }
+    }
+    if (Number.isFinite(pid) && pid !== null && pid > 0) {
+      try { process.kill(pid, "SIGTERM"); } catch { /* already dead */ }
+    }
+    try { unlinkSync(this.socketPath); } catch { /* not present */ }
+    try { unlinkSync(this.pidPath); } catch { /* not present */ }
+  }
```
[the remark this answers is not in the document]
Comment (efenocchi) [reply src/embeddings/client.ts:313]: Not actionable — this was flagged from a stale coverage snapshot. The current CI run shows `src/embeddings/client.ts` at 93.51% statements / 82.71% branches / 96.29% functions / 94.33% lines, all above the configured per-file thresholds (90/80/90/90 in `vitest.config.ts`). Re-checking the latest CI on this branch should clear it.

File: tests/claude-code/session-start-setup-hook.test.ts [hunk 16]
```diff
@@ -227,11 +227,11 @@ describe("session-start-setup hook — embed daemon warmup", () => {
...
 
-  it("skips warmup when the master HIVEMIND_EMBEDDINGS=false flag is set", async () => {
-    await runHook({ HIVEMIND_EMBEDDINGS: "false" });
+  it("skips warmup when the user has disabled embeddings in config", async () => {
+    await runHook({ EMBEDDINGS_DISABLED_FOR_TEST: "1" });
     expect(embedWarmupMock).not.toHaveBeenCalled();
     expect(debugLogMock).toHaveBeenCalledWith(
-      "embed daemon warmup skipped: HIVEMIND_EMBEDDINGS=false",
+      "embed daemon warmup skipped: embeddings disabled in ~/.deeplake/config.json (run `hivemind embeddings enable` to opt in)",
     );
```
[the remark this answers is not in the document]
Comment (efenocchi) [reply tests/claude-code/session-start-setup-hook.test.ts:260]: Acknowledged — will follow up in a small test-hygiene cleanup PR alongside #17.

File: tests/hermes/hermes-capture-hook.test.ts [hunk 17]
```diff
@@ -265,14 +265,20 @@ describe("hermes capture hook — message_embedding column", () => {
...
       extra: { prompt: "disabled" },
     });
-    await runHook({ HIVEMIND_EMBEDDINGS: "false" });
+    const { writeFileSync, mkdtempSync } = await import("node:fs");
+    const { tmpdir } = await import("node:os");
+    const { join } = await import("node:path");
+    const dir = mkdtempSync(join(tmpdir(), "hermes-cap-disabled-"));
+    const cfgPath = join(dir, "config.json");
+    writeFileSync(cfgPath, JSON.stringify({ embeddings: { enabled: false } }), "utf-8");
+    await runHook({ HIVEMIND_CONFIG_PATH: cfgPath });
```
[the remark this answers is not in the document]
Comment (efenocchi) [reply tests/hermes/hermes-capture-hook.test.ts:313]: Acknowledged — will follow up in the same test-hygiene cleanup PR as #16.

File: tests/claude-code/session-start-setup-hook.test.ts (hunk 16 above)
[the remark this answers is not in the document]
Comment (efenocchi) [reply tests/claude-code/session-start-setup-hook.test.ts:260]: Fixed in 9b57aaf — `runHook()` now snapshots the original value of every touched env key on first mutation, and afterEach restores them all. No leakage into later tests in the same vitest worker.

File: tests/hermes/hermes-capture-hook.test.ts (hunk 17 above)
[the remark this answers is not in the document]
Comment (efenocchi) [reply tests/hermes/hermes-capture-hook.test.ts:313]: Fixed in 9b57aaf — same pattern as #16: `runHook()` snapshots `HIVEMIND_CONFIG_PATH` (and `HIVEMIND_CAPTURE`) on first mutation; afterEach restores them and rmSync's the tmp config dir the user-disabled test creates.

File: embeddings/embed-daemon.js [hunk 18]
```diff
@@ -216,6 +258,15 @@ var EmbedDaemon = class {
     }
   }
   async dispatch(req) {
+    if (req.op === "hello") {
```
Comment (kaghni) [inline embeddings/embed-daemon.js:263]: @coderabbitai what hello stands for, here, why is hardcoded?

File: src/cli/embeddings.ts [hunk 19]
```diff
@@ -131,34 +134,77 @@ function ensureSharedDeps(): void {
...
   const link = join(install.pluginDir, "node_modules");
+  // Don't try to overwrite a real `node_modules` directory: `symlinkForce`
+  // calls `unlinkSync` first, which throws EISDIR on directories and would
+  // abort `hivemind embeddings install` partway through, leaving some
+  // agents linked and others not. Defer to whatever the user/marketplace
+  // installed there — the same state `status` already surfaces as
+  // `owns-own-node-modules`. (Symlinks at this path, including stale ones
+  // pointing at a defunct shared-deps target, still go through
+  // `symlinkForce` so we replace them with the canonical target.)
+  const state = linkStateFor(install);
```
Comment (kaghni) [inline src/cli/embeddings.ts:151]: @coderabbitai, why is there even a problem of overriding node_modules?

File: src/embeddings/daemon.ts
Comment (kaghni) [inline src/embeddings/daemon.ts:1]: @coderabbitai what daemon.ts is for and where is it used? briefily

File: src/embeddings/nomic.ts
Comment (kaghni) [inline src/embeddings/nomic.ts:1]: @coderabbitai same here, brief usage and description of nomic.ts

File: src/embeddings/protocol.ts
Comment (kaghni) [inline src/embeddings/protocol.ts:1]: @coderabbitai how protocol works for skillify?

**Comment by kaghni** [comment]
@efenocchi, Pi and OpenClaw have no changes, but I'm not sure they should.

**Comment by efenocchi** [comment]
@kaghni Confirmed correct on both — verified the surface:

**Pi**: `pi/extension-source/hivemind.ts` ships as raw TS with a 30-LOC `tryEmbedOverSocket` that fails silent (no `EmbeddingsClient`, no `enqueueNotification`, no hello-handshake, no recycle) — none of the surfaces this PR touches exist client-side. The one pi bundle rebuilt (`pi/bundle/wiki-worker.js`) is the shared wiki-worker, unrelated. Pi *does* pick up the canonical daemon fix (`embeddings/embed-daemon.js`: +55/-4) on the next `hivemind embeddings install`.

**OpenClaw**: MCP-mode plugin with `hivemind_*` tool contracts only. It currently **doesn't produce embeddings at all** — every write explicitly NULLs the vector columns:
- `upsertRowSql`: `summary_embedding = NULL` on UPDATE, `NULL` literal on INSERT
- session capture: `message_embedding` omitted from the column list → defaults NULL

It only **consumes** embeddings (semantic search takes `queryEmbedding` as an input parameter and filters `WHERE ARRAY_LENGTH(summary_embedding, 1) > 0`, i.e. rows written by other agents). So there's no surface for this PR to touch.

**Follow-up PR — tracked separately, not in this one.** We're going to:
1. Fix a real bug in pi: the comment at `pi/extension-source/hivemind.ts:166-175` promises auto-spawn-on-miss, but the implementation only calls `connect()` and settles \`null\` on error. So pi silently degrades to NULL embeddings any time it's the *first* agent to talk to the daemon on a given box. That's a regression vector identical to what this PR is closing — just on a different code path.
2. Make OpenClaw an embedding producer: ~60 LOC of \`tryEmbedOverSocket\` + spawn-on-miss, wired into the two write sites. Talks to the same \`/tmp/hivemind-embed-<uid>.sock\` daemon the other agents already spawn.

**Why a separate PR rather than tacking it onto this one**:
- **Scope**: this PR is "*fix* silent NULL embeddings after marketplace upgrades" — restoring producers that should have worked. Adding a new producer (openclaw) is a feature, not a fix; bundling them obscures the bisect/revert story if either side regresses.
- **Opt-in invariant**: this PR established (\`src/user-config.ts\`) that embeddings are **explicit opt-in** — env unset → \`enabled: false\`, no 600MB download by surprise. OpenClaw becoming a producer needs to respect that: it can only embed if \`~/.hivemind/embed-deps/embed-daemon.js\` is already present (deposited by \`hivemind embeddings install\` via *some* other agent). That decision and its NULL-fallback behavior deserve their own review surface, not a buried sub-commit.
- **Edge case coverage**: spawn-on-miss has 11 distinct states I want covered by tests (binary-missing, stale-socket, pidfile-points-to-dead-PID, two-process spawn race, pidfile-alive-but-socket-gone, etc.). Those tests double the test footprint of this PR — cleaner to land them alongside the producer code they exercise.

I'll open the follow-up against \`main\` once #168 is merged.

**Review by kaghni — APPROVED** [review]
@
